lib/src/Uncertainty/Distribution/EmpiricalBernsteinCopula.cxx

```diff
@@ -0,0 +1,422 @@
+  for (UnsignedInteger i = 0; i < dimension; ++i)
+    if ((point[i] <= 0.0) || (point[i] >= 1.0)) return -SpecFunc::LogMaxScalar;
+  Scalar sumPDFValue = 0.0;
+  Point logX(dimension);
+  Point log1mX(dimension);
+  for (UnsignedInteger i = 0; i < dimension; ++i)
+  {
+    logX[i] = std::log(point[i]);
+    log1mX[i] = log1p(-point[i]);
+  }
+  const UnsignedInteger size = copulaSample_.getSize();
+  UnsignedInteger linearIndex = 0;
+  Scalar maxLogPDF = 0.0;
+  const Scalar logW = -std::log(1.0 * size);
+  for (UnsignedInteger i = 0; i < size; ++i)
+  {
+    Scalar logPDFAtom = 0.0;
+    for (UnsignedInteger j = 0; j < dimension; ++j)
+    {
+      logPDFAtom += (logFactors_[linearIndex] - 1.0) * logX[j] + (binNumber_ - logFactors_[linearIndex]) * log1mX[j] + logW - logBetaFactors_[linearIndex];
+      ++linearIndex;
+    } // j
+    sumPDFValue += std::exp(logPDFAtom);
+  } // i
+  return std::log(sumPDFValue);
+}
+
+/* Get the CDF of the EmpiricalBernsteinCopula */
+Scalar EmpiricalBernsteinCopula::computeCDF(const Point & point) const
+{
+  const UnsignedInteger dimension = getDimension();
+  if (point.getDimension() != dimension) throw InvalidArgumentException(HERE) << "Error: the given point must have dimension=" << dimension << ", here dimension=" << point.getDimension();
+  Scalar cdfValue = 0.0;
+  const UnsignedInteger size = copulaSample_.getSize();
+  UnsignedInteger linearIndex = 0;
+  for (UnsignedInteger i = 0; i < size; ++i)
+  {
+    Scalar cdfAtom = 1.0;
+    for (UnsignedInteger j = 0; j < dimension; ++j)
+    {
+      cdfAtom *= SpecFunc::RegularizedIncompleteBeta(logFactors_[linearIndex], binNumber_ - logFactors_[linearIndex] + 1.0, point[j]);
+      ++linearIndex;
+    } // j
+    cdfValue += cdfAtom;
+  } // i
+  return cdfValue / size;
+}
+
+/* Compute the probability content of an interval */
+Scalar EmpiricalBernsteinCopula::computeProbability(const Interval & interval) const
+{
+  if (interval.isNumericallyEmpty()) return 0.0;
+  const UnsignedInteger dimension = getDimension();
+  Scalar probabilityValue = 0.0;
+  const UnsignedInteger size = copulaSample_.getSize();
+  UnsignedInteger linearIndex = 0;
+  const Point lower(interval.getLowerBound());
+  const Point upper(interval.getUpperBound());
+  for (UnsignedInteger i = 0; i < size; ++i)
+  {
+    Scalar probabilityAtom = 1.0;
+    for (UnsignedInteger j = 0; j < dimension; ++j)
+    {
+      const Scalar r = logFactors_[linearIndex];
+      const Scalar s = binNumber_ - logFactors_[linearIndex] + 1.0;
+      probabilityAtom *= SpecFunc::RegularizedIncompleteBeta(r, s, upper[j]) - SpecFunc::RegularizedIncompleteBeta(r, s, lower[j]);
+      ++linearIndex;
+    } // j
+    probabilityValue += probabilityAtom;
+  } // i
+  return probabilityValue / size;
+}
+
+/* Get the distribution of the marginal distribution corresponding to indices dimensions */
+Distribution EmpiricalBernsteinCopula::getMarginal(const Indices & indices) const
+{
```

Comment:
Why don't you just `return new EmpiricalBernsteinCopula(copulaSample_.getMarginal(indices), binNumber_, true);`?

Reply:
@dbarbier It is to avoid a call to update(), which is not so cheap

Reply:
Okay